clippy_lints/src/needless_question_mark.rs

```diff
@@ -0,0 +1,191 @@
+use rustc_errors::Applicability;
+use rustc_hir::def::{CtorKind, CtorOf, DefKind, Res};
+use rustc_hir::{Body, Expr, ExprKind, LangItem, MatchSource, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::DefIdTree;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+use crate::utils;
+use if_chain::if_chain;
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// Suggest for terminating expressions
+    /// ```rust,ignore
+    /// value.inner_option;
+    /// value.inner_result;
+    /// ```
+    /// instead of
+    /// ```rust,ignore
+    /// Some(value.inner_option?);
+    /// Ok(value.inner_result?);
+    /// ```
+    ///
+    /// **Why is this bad?** There's no reason to use ? to short-circuit when execution of the body will end there anyway.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // struct Outer {
+    /// //     inner: Option<String>
+    /// // }
+    /// //
+    /// // let a = Some(Outer { inner: Some("hello".into()) });
+    /// // a.and_then(|a| Some(a.inner?))
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // struct Outer {
+    /// //     inner: Option<String>
+    /// // }
+    /// //
+    /// // let a = Some(Outer { inner: Some("hello".into()) });
+    /// // a.and_then(|a| a.inner)
```

Comment:
Don't comment out doc tests. You should make those tests pass. If you want to define things that won't appear in the doc output you can prefix it with `#`. E.g. `# let a = Some(..)` will not show in the documentation.